### ISSUE TITLE
1.8.15 regresssion - build with different outputs on diffrent arches

### ISSUE THREAD

**Issue by ngothan** [doxygen/doxygen#6838]
Description of problem:
https://koji.fedoraproject.org/koji/taskinfo?taskID=32791130

The build fails with an error about the noarch package building different outputs on different arches.

BuildError: The following noarch package built differently on different architectures: dpdk-doc-17.11.2-6.fc30.noarch.rpm
rpmdiff output was:
added       /usr/share/doc/dpdk/api/dir_16a7fac6d3a9be16fa23e31b05d0f6a0.html
added       /usr/share/doc/dpdk/api/dir_274a49b40ed917259b7b614d6d0cfbf7.html
added       /usr/share/doc/dpdk/api/dir_2841d85298525af6192dea3e61ef688d.html
removed     /usr/share/doc/dpdk/api/dir_acbd5fab9c0cd486dd354deb15c43b30.html
removed     /usr/share/doc/dpdk/api/dir_ae18a39335608f73cd855a33d4d53394.html
removed     /usr/share/doc/dpdk/api/dir_ca540074b6ef21b68821b2fb54f8c10f.html

Version-Release number of selected component (if applicable):
doxygen-1.8.15

How reproducible:
always

Steps to Reproduce:
1.build dpdk in fedora rawhide

Actual results:
build fails due to discrepancies in doxygen output for docs noarch package

Expected results:
doxygen produces the same results on all arches

Note: it works fine with doxygen-1.8.14

**Comment by albert-github** [comment]
Can you indicate what builds differently between the different architectures?
Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.

**Comment by lrineau** [comment]
The diff seems to be files that have different names:
> ```
> rpmdiff output was:
> added       /usr/share/doc/dpdk/api/dir_16a7fac6d3a9be16fa23e31b05d0f6a0.html
> added       /usr/share/doc/dpdk/api/dir_274a49b40ed917259b7b614d6d0cfbf7.html
> added       /usr/share/doc/dpdk/api/dir_2841d85298525af6192dea3e61ef688d.html
> removed     /usr/share/doc/dpdk/api/dir_acbd5fab9c0cd486dd354deb15c43b30.html
> removed     /usr/share/doc/dpdk/api/dir_ae18a39335608f73cd855a33d4d53394.html
> removed     /usr/share/doc/dpdk/api/dir_ca540074b6ef21b68821b2fb54f8c10f.html
> ```

Do you see something in the generation of the hash that could vary from one platform to another?

**Comment by albert-github** [comment]
The hash generation on the different platforms should identical, but without an example it is hard to tell where the problem lies. looking at the doxygen code in the file `dirdef.cpp` the hash is determined in the function `getOutputFileBase` by means of  `encodeDirName` which is also in `dirdef.cpp` that has the directory name as input. The MD5 determination is done by standard routines that are also provided by doxygen(directory `libmd5`).

The only recent change I see is in `md5.c` where:
`    register UWORD32 a, b, c, d;`
is replaced by
`    UWORD32 a, b, c, d;`
but this shoul have no effect, and the `register` keyword is deprecated since C++11.

Best way to debug would be to place some print statements in `getOutputFileBase` to see what here the input and output is on the different platforms.

**Comment by ngothan** [comment]
The following commit im master branch causes the reported issue. I have reverted the change and the problem is gone. 

```
commit 5d79d65df3b66554c8e9630fd3bea322c3e36f0d
Author: Christoph Lipka <c-lipka@users.noreply.github.com>
Date:   Sat Apr 14 11:37:58 2018 +0200

    Include "empty" directories in the documentation if they contain a `.dox` file (or similar) documenting the directory itself.

diff --git a/src/dirdef.cpp b/src/dirdef.cpp
index b247ba74..6631ed7f 100644
--- a/src/dirdef.cpp
+++ b/src/dirdef.cpp
@@ -930,14 +930,14 @@ void buildDirectories()
     for (;(fd=fni.current());++fni)
     {
       //printf("buildDirectories %s\n",fd->name().data());
-      if (fd->getReference().isEmpty() && !fd->isDocumentationFile())
+      if (fd->getReference().isEmpty())
       {
         DirDef *dir;
         if ((dir=Doxygen::directories->find(fd->getPath()))==0) // new directory
         {
           dir = DirDef::mergeDirectoryInTree(fd->getPath());
         }
-        if (dir) dir->addFile(fd);
+        if (dir && !fd->isDocumentationFile()) dir->addFile(fd);
       }
       else
       {
```

**Comment by albert-github** [comment]
@c-lipka can you comment on the changes?

**Comment by c-lipka** [comment]
The comment to the corresponding pull request is pretty much all I can offer.

The content of the diff'ing files in question might give a clue: If they happen to be documentations of directories not included by earlier versions of Doxygen, then a connection to that commit seems highly plausible.

Could it be that the code generating the MD5 hashes isn't triggered for those files, leaving the hash value uninitialized?

**Comment by albert-github** [comment]
@c-lipka thanks, without the source code it is hard to tell.
@ngothan 
- which version of dkdp did you use?
- where did you get the source code (so we are talking about the same) and we can try to reproduce the issue and see what the exact cause of the problem is.

**Comment by albert-github** [comment]
@ngothan 
I did some tests (with dpkp version 19.02), but I was not able to reproduce the problem (I only have Windows and Cygwin).
I didn't even see any of the filenames as indicated in the issue report.
- are these 6 lines the only 6 lines regarding `add` / `remove` or are there more (just to have an indication of the possible problem)?

It would be extremely useful to have:
- The source code you used (on the arches where you saw the differences) as well as the complete HTML output (but without executables, object files and libraries that might have been build as well).
- the used Doxyfiles (so with a.o. the substituted values for `@TOPDIR@`, `@API_EXAMPLES@`, `@STRIP_FROM_PATH@`) for both calculations.
- the output when you un-comment the debug statements in  the function `DirDef::getOutputFileBase` in `dirdef.cpp`.

**Comment by ngothan** [comment]
@albert-github 
i assume you may have used only doxygen built for x86_64 or i686 platform in the test. To produce the problem you have to build doxygen explicitly for both arch (x86_64 and i686) for the test. And when testing, please use both. The problem may not occur with Windows and Cygwin.

There are the only 6 lines regarding add / remove which causes the build failure in our fedora  build system. For more infos i will post later today.

Thanks

**Comment by ngothan** [comment]
@albert-github 
I found out the problem is the name of the output directory. In our case the name of the output directory is different on i686 and x86_84 platform. 

To reproduce the problem i made a testcase. It generates the HTML output under different output directories (a and b) and then compares the result between those two directories. (Please look at the log file in doxygen-testcase.tar.gz)

diff -sr a b | grep -v identical
Only in b/doc/html/api: dir_1c6b3428ba6da23642a8fd43c0d27cb9.html
Only in a/doc/html/api: dir_1fcc60d3b46ec2cb829cd268fbf0b995.html
Only in b/doc/html/api: dir_22466a4aca50b69c94ae3a61afd7f4ac.html
Only in a/doc/html/api: dir_56985c2299c680a23218e62ee374bdce.html
Only in b/doc/html/api: dir_84aafc86c89ed9cb92fc605e636e3134.html
Only in a/doc/html/api: dir_a8a6ed2d9486ef56f6bb32e0b8fb740e.html

I upload the testcase which includes the source code, HTML output, doxygen output and the Doxyfiles doc/api/doxy-api.conf

[doxygen-testcase.tar.gz](https://github.com/doxygen/doxygen/files/2879645/doxygen-testcase.tar.gz)

**Comment by albert-github** [comment]
Thanks for the case

I've tried to reproduce the problem but I didn't succeed yet.
When running I get an number of warnings starting with:
```
warning: tag INPUT: input source `drivers/crypto/scheduler' does not exist
warning: tag INPUT: input source `drivers/net/bnxt' does not exist
warning: tag INPUT: input source `drivers/net/bonding' does not exist
warning: tag INPUT: input source `drivers/net/i40e' does not exist
warning: tag INPUT: input source `drivers/net/ixgbe' does not exist
warning: tag INPUT: input source `drivers/net/softnic' does not exist
warning: source drivers/crypto/scheduler is not a readable file or directory... skipping.
warning: source drivers/net/bnxt is not a readable file or directory... skipping.
warning: source drivers/net/bonding is not a readable file or directory... skipping.
warning: source drivers/net/i40e is not a readable file or directory... skipping.
warning: source drivers/net/ixgbe is not a readable file or directory... skipping.
warning: source drivers/net/softnic is not a readable file or directory... skipping.
```
and some subsequent warnings, I don't think they are a problem.

Can you replace in `mk/rte.sdkdoc.mk` (line 83):
`             doxygen -`
with
`             doxygen -x -`

and attach the output to this issue?

**Comment by ngothan** [comment]
@albert-github 

Attach the output with doxygen -x
[log1.gz](https://github.com/doxygen/doxygen/files/2884736/log1.gz)

**Comment by albert-github** [comment]
I found the problematic part:
```
61c61
<                          /home/than/doxygen-testcase/a/doc/html/examples.dox
---
>                          /home/than/doxygen-testcase/b/doc/html/examples.dox
```
This is adding a generated file to the INPUT directive, but the file is located in (system depending)  html output directory.

I added  `echo STRIP_FROM_PATH=$(RTE_SDK) $(RTE_OUTPUT)/doc/html` (actually an analogous statement suitable for my environment) to the `mk/rte.sdkdoc.mk` and this did the trick.

A better solution would, of course, be that the generated files lie in a system independent directory.
